File: public/javascripts/gildedRose.js
function Item(name, sell_in, quality, update) {
    this.name = name;
    this.sell_in = sell_in;
    this.quality = quality;
    this.update = update;
}

function GildedRose() {
    this.items = [];
    this.init();
}

GildedRose.prototype.init = function () {
    this.items.push(new Item('+5 Dexterity Vest', 10, 20, Item.prototype.update_by_default));
    this.items.push(new Item('Aged Brie', 2, 0, Item.prototype.update_agedBrie));
    this.items.push(new Item('Elixir of the Mongoose', 5, 7, Item.prototype.update_by_default));
    this.items.push(new Item('Sulfuras, Hand of Ragnaros', 0, 80, Item.prototype.update_Sulfuras));
    this.items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20, Item.prototype.update_BackstagePasses));
    this.items.push(new Item('Conjured Mana Cake', 3, 6, Item.prototype.update_by_default));
}


GildedRose.prototype.update_quality = function () {
    $.each(this.items, function(index, item) { item.update() });
}


Item.prototype.update_agedBrie = function() {
        if (this.quality < 50) {
            this.quality = this.quality + 1
        }
        this.sell_in = this.sell_in - 1;
        if (this.sell_in < 0) {
            if (this.quality < 50) {
                this.quality = this.quality + 1
            }
        }
}

Item.prototype.update_BackstagePasses = function() {
    if (this.quality < 50) {
        this.quality = this.quality + 1
        if (this.sell_in < 11) {
            if (this.quality < 50) {
                this.quality = this.quality + 1
            }
        }
        if (this.sell_in < 6) {
            if (this.quality < 50) {
                this.quality = this.quality + 1
            }
        }
    }
    this.sell_in = this.sell_in - 1;
    if (this.sell_in < 0) {
        this.quality = this.quality - this.quality
    }
}

Item.prototype.update_Sulfuras = function() {

}

Item.prototype.update_by_default = function() {
    if (this.quality > 0) {
        this.quality = this.quality - 1
    }
    this.sell_in = this.sell_in - 1;
    if (this.sell_in < 0) {
        if (this.quality > 0) {
            this.quality = this.quality - 1
        }
    }
}

File: app/models/GildedRose.java
package models;

import java.util.ArrayList;
import java.util.List;


public class GildedRose {

    private final List<Item> items = new ArrayList<Item>();

    /**
     * @param args
     */
    public static void main(String[] args) {
 
        System.out.println("OMGHAI!");
        GildedRose inn = new GildedRose();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        inn.updateQuality();
        System.out.println(inn);
    }

    public GildedRose() {
        items.add(new Item("+5 Dexterity Vest", 10, 20));
        items.add(new Item("Aged Brie", 2, 0));
        items.add(new Item("Elixir of the Mongoose", 5, 7));
        items.add(new Item("Sulfuras, Hand of Ragnaros", 0, 80));
        items.add(new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20));
        items.add(new Item("Conjured Mana Cake", 3, 6));
    }

    public List<Item> getItems() {
        return items;
    }

    /**
     * Votre mission, si vous l'acceptez...
     * <p/>
     * Ajouter un item magique avec le nom "Toujours A Sang"
     * qui, quoi qu'il arrive, aura toujours une qualité à 100
     */
    public void updateQuality() {
        for (int i = 0; i < items.size(); i++) {
            if ((!"Aged Brie".equals(items.get(i).getName())) && !"Backstage passes to a TAFKAL80ETC concert".equals(items.get(i).getName())) {
                if (items.get(i).getQuality() > 0) {
                    if (!"Sulfuras, Hand of Ragnaros".equals(items.get(i).getName())) {
                        items.get(i).setQuality(items.get(i).getQuality() - 1);
                    }
                }
            } else {
                if (items.get(i).getQuality() < 50) {
                    items.get(i).setQuality(items.get(i).getQuality() + 1);

                    if ("Backstage passes to a TAFKAL80ETC concert".equals(items.get(i).getName())) {
                        if (items.get(i).getSellIn() < 11) {
                            if (items.get(i).getQuality() < 50) {
                                items.get(i).setQuality(items.get(i).getQuality() + 1);
                            }
                        }

                        if (items.get(i).getSellIn() < 6) {
                            if (items.get(i).getQuality() < 50) {
                                items.get(i).setQuality(items.get(i).getQuality() + 1);
                            }
                        }
                    }
                }
            }

            if (!"Sulfuras, Hand of Ragnaros".equals(items.get(i).getName())) {
                items.get(i).setSellIn(items.get(i).getSellIn() - 1);
            }

            if (items.get(i).getSellIn() < 0) {
                if (!"Aged Brie".equals(items.get(i).getName())) {
                    if (!"Backstage passes to a TAFKAL80ETC concert".equals(items.get(i).getName())) {
                        if (items.get(i).getQuality() > 0) {
                            if (!"Sulfuras, Hand of Ragnaros".equals(items.get(i).getName())) {
                                items.get(i).setQuality(items.get(i).getQuality() - 1);
                            }
                        }
                    } else {
                        items.get(i).setQuality(items.get(i).getQuality() - items.get(i).getQuality());
                    }
                } else {
                    if (items.get(i).getQuality() < 50) {
                        items.get(i).setQuality(items.get(i).getQuality() + 1);
                    }
                }
            }
        }
    }

    @Override
    public String toString() {
        return "GildedRose{" +
                "items=\n" + items +
                '}';
    }
}